### PR TITLE
Feature/dan branch support

### PR DESCRIPTION
**Pull request by qualidan**

### CODE REVIEW

File: package/cloudshell/iac/terraform/services/tf_proc_exec.py [hunk 1]
```diff
@@ -227,9 +227,8 @@ def _set_service_status(self, status: str, description: str):
             description
         )
 
-    def _init_backend_config(self) -> str:
-        backend_attribute_name = \
-            f"{self._shell_helper.tf_service.cloudshell_model_name}.{ATTRIBUTE_NAMES.REMOTE_STATE_PROVIDER}"
+    def _init_backend_config(self):
+        backend_attribute_name = f"{self._shell_helper.tf_service.cloudshell_model_name}.Remote State Provider"
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/services/tf_proc_exec.py]: why did you remove the use of constant for the attribute name?

File: package/cloudshell/iac/terraform/services/svc_attribute_handler.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        self._tf_service = tf_service
+
+    def get_attribute(self, attribute_name: str):
+        services = self._api.GetReservationDetails(self._sandbox_id).ReservationDescription.Services
+        for service in services:
+            if service.Alias == self._tf_service.name:
+                for attribute in service.Attributes:
+                    if attribute.Name == f"{self._tf_service.cloudshell_model_name}.{attribute_name}":
+                        return attribute.Value
+        return ""
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/services/svc_attribute_handler.py]: this code doesnt makes sense to me. I think it also conflicts with the refactoring that @Chris530906 is doing

File: package/cloudshell/iac/terraform/downloaders/downloader.py [hunk 3]
```diff
@@ -9,15 +10,20 @@ def __init__(self, shell_helper: ShellHelperObject):
...
         url = self._shell_helper.tf_service.github_terraform_module_url
-        if hasattr(self._shell_helper.tf_service, 'github_token'):
-            token = self._shell_helper.api.DecryptPassword(self._shell_helper.tf_service.github_token).Value
-        else:
-            token = ""
+        attr_handler = ServiceAttrHandler(
+            self._shell_helper.api,
+            self._shell_helper.sandbox_id,
+            self._shell_helper.tf_service
+        )
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/downloaders/downloader.py]: this should not be initialized here

File: package/cloudshell/iac/terraform/downloaders/downloader.py [hunk 4]
```diff
@@ -9,15 +10,20 @@ def __init__(self, shell_helper: ShellHelperObject):
...
+
+        token = attr_handler.get_attribute('github_token')
+        branch = attr_handler.get_attribute('branch')
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/downloaders/downloader.py]: why not use the attribute name constants?

**Review by alexazarh — CHANGES_REQUESTED** [review]

File: package/cloudshell/iac/terraform/services/tf_proc_exec.py (hunk 1 above)
Comment (qualidan) [reply package/cloudshell/iac/terraform/services/tf_proc_exec.py]: don`t understand how it got to this state...I was the one who defined this :ATTRIBUTE_NAMES.REMOTE_STATE_PROVIDER

File: package/cloudshell/iac/terraform/services/svc_attribute_handler.py (hunk 2 above)
Comment (qualidan) [reply package/cloudshell/iac/terraform/services/svc_attribute_handler.py]: I talked to Chris and can explain why this is needed

File: package/cloudshell/iac/terraform/downloaders/downloader.py (hunk 4 above)
Comment (qualidan) [reply package/cloudshell/iac/terraform/downloaders/downloader.py]: correct...will fix

File: package/cloudshell/iac/terraform/services/svc_attribute_handler.py (hunk 2 above)
Comment (alexazarh) [reply package/cloudshell/iac/terraform/services/svc_attribute_handler.py]: Lets discuss tomorrow about this. But generally speaking, it should be done only once. Not for every attribute that we want to access.

File: package/cloudshell/iac/terraform/services/tf_proc_exec.py (hunk 1 above)
Comment (qualidan) [reply package/cloudshell/iac/terraform/services/tf_proc_exec.py]: fix in commit 16b09f2f1bc270ba7499e05e48d37dee3a4d57e6

File: package/cloudshell/iac/terraform/services/svc_attribute_handler.py (hunk 2 above)
Comment (qualidan) [reply package/cloudshell/iac/terraform/services/svc_attribute_handler.py]: fixed using a static attr list

File: package/cloudshell/iac/terraform/downloaders/downloader.py (hunk 3 above)
Comment (qualidan) [reply package/cloudshell/iac/terraform/downloaders/downloader.py]: fixed (moved to shell helper)

File: package/cloudshell/iac/terraform/models/shell_helper.py [hunk 5]
```diff
@@ -16,3 +17,9 @@ def __init__(self, api: CloudShellAPISession, res_id: str, tf_service: any, logg
         self.logger = logger
         self.sandbox_messages = sandbox_messages
         self.live_status_updater = live_status_updater
+
+        self.attr_handler = ServiceAttrHandler(
+            self.api,
+            self.sandbox_id,
+            self.tf_service
+        )
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/models/shell_helper.py]: its an anti-pattern to initialize services here. It should be provided as dependency injection.

File: package/cloudshell/iac/terraform/services/tf_proc_exec.py [hunk 6]
```diff
@@ -227,7 +227,7 @@ def _set_service_status(self, status: str, description: str):
             description
         )
 
-    def _init_backend_config(self) -> str:
+    def _init_backend_config(self):
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/services/tf_proc_exec.py]: why?

File: package/cloudshell/iac/terraform/services/svc_attribute_handler.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+    def _populate_attr_list(self):
+        services = self._api.GetReservationDetails(self._sandbox_id).ReservationDescription.Services
+        for service in services:
+            if service.Alias == self._tf_service.name:
+                for attribute in service.Attributes:
+                    if "." in attribute.Name:
+                        self._attr_list.append(attribute.Name.split(".")[1])
+                    else:
+                        self._attr_list.append(attribute.Name)
```
Comment (alexazarh) [inline package/cloudshell/iac/terraform/services/svc_attribute_handler.py]: lets cache also the values

**Review by alexazarh — CHANGES_REQUESTED** [review]
